### ISSUE TITLE
Improve handling of Java candidate

### ISSUE THREAD

**Issue by vpavic** [sdkman/sdkman-cli#696]
As discussed on Gitter in [sdkman/user-issues](https://gitter.im/sdkman/user-issues?at=5ca86bf3bd70a40d5f047d46) channel, it would be nice if SDKMAN could provide some special handling for Java candidates in order to address some of the complexity that comes from presence of different JDK flavors, as well as different kinds of _current_ releases (LTS vs non-LTS).

For instance, I have a preference for Zulu JDKs and generally don't need/want JDKs from other vendors. SDKMAN allows to install individual Zulu JDKs and use `default` command to set the local default, however when using `upgrade` command SDKMAN will offer to install candidate that is marked as default on the remote - so even though I have the same or even newer release from a different vendor (note that at the time of writing this, Java candidate does not seem to have any candidate marked as default on the remote).

One possible solution would be to have different candidates for each JDK vendor (e.g. `java-open`, `java-zulu`, `java-adpt-hs`, `java-adpt-j9`, etc.). This would perhaps be more natural as for other candidates (to my knowledge) there is a 1:1 mapping between the candidate and the vendor, whereas for Java it is 1:n due to multiple vendors contributing to Java candidate. This would however add complexity in other places, as it's unclear how `JAVA_HOME` would behave when there are multiple Java candidates installed.

**Comment by vpavic** [comment]
The expectations could be expressed as:

- given I have a JDK from vendor X installed, when `upgrade` command is used SDKMAN should offer to install a new release from vendor X (if exists)
- given I have a JDK from vendor X installed, when `upgrade` command is used SDKMAN shouldn't offer to install a release from vendor Y (or any other vendor that isn't X)

**Comment by marc0der** [comment]
I think all the above observations are very valid, but are limited to a single facet (upgrading) of the scope of the improvements I was intending to make for the Java candidate. My thoughts are that in the case of Java, _some animals are more equal than others_ on sdkman :smile:. With that in mind, I would like to let JDKs work a bit differently than other SDKs, granting some concessions to how sdkman handles JDKs. 

Here are some of the ideas I had:

- provide a custom list view for java versions on `sdk ls java`, perhaps having it's own template rendered on the serverside with neat columns and perhaps sections per vendor.
- introduce a JDKs page to the website (will have details about each vendor, current LTS version etc)
- extend the base commands to include a vendor qualifier, so allowing something like `sdk install java [vendor] [version]`. it would also work for the `set` and `default` commands. we might need to think this through a bit more.
- as mentioned above, we could introduce a notion of _pinning_ a JDK vendor, so that upgrades would be applied for a given vendor only. This could be done by using the `default` command or even introducing a brand new top-level command called `pin` that would only work for Java.

Regarding introducing candidates for each java/vendor combination would break a lot of things and I think pollute our domain somewhat. I would prefer keeping _candidate_ pure and introducing a new domain concept for _vendor_ instead. We already pushed the limit by overloading _version_ to include the _vendor_, but shifting it to _candidate_ would be equally wrong in my opinion. If we gave vendor it's own domain concept, it would also make the above functionality a lot easier to implement. wdyt?

**Comment by kevin-stuart** [comment]
Just a thought: Maybe just adding vendor could not be enough for the future and this is an opportunity to make things more future-proof. You now only add one new dimension "vendor". But maybe in the future, there are more fine-grained splits like multiple GraalVM vendors.

I would suggest thinking about something hierarchical like a file system, which has served Unix very well. For example, you could an "oracle" flavor and add have OpenJDK and GraalVM as possible children. The grouping could also be different to have the type like OpenJDK first and then the vendor. Or you could support both. It could also be used to group variants of SDKs better, for example for Zulu you also have a version with JavaFX or you have versions with different GCs like j9 or Hotspot you might want to use.

I know this sounds complicated and this definitely needs more thought, but I just wanted to throw out this idea and start some brainstorming. If I say `sdk install java/openjdk/zulu/fx version` and would like to pin the variant java/openjdk/zulu/fx I would only get new releases for Zulu with JavaFX. And I might want to have a parallel install of GraalVM from Oracle, which I also want to keep up to date.

I think it might be important to think about GraalVM now if it exists in parallel with OpenJDK now, the other option could be to create a new variant graalvm instead of java, but the downside is that GraalVM clashes with `java` in `PATH`, so the current approach makes sense.

**Comment by marc0der** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Improve handling of Java candidate](https://trello.com/c/Ah6VoDV9/70-improve-handling-of-java-candidate)

**Comment by marc0der** [comment]
I've fixed some bugs in the Java list view around displaying of _Unknown_ vendors (locally installed versions that don't belong to a specific JDK vendor).

**Comment by vpavic** [comment]
I was about to ping you @marc0der about displaying local installs, so thanks for taking a look at that. However, latest changes have broken the listing for me:

```bash
$ sdk ls java


<!DOCTYPE html>
<html lang="en">
    <head>
        <title>Error</title>
        <link rel="shortcut icon" href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAlFJREFUeNqUU8tOFEEUPVVdNV3dPe8xYRBnjGhmBgKjKzCIiQvBoIaNbly5Z+PSv3Aj7DSiP2B0rwkLGVdGgxITSCRIJGSMEQWZR3eVt5sEFBgTb/dN1yvnnHtPNTPG4PqdHgCMXnPRSZrpSuH8vUJu4DE4rYHDGAZDX62BZttHqTiIayM3gGiXQsgYLEvATaqxU+dy1U13YXapXptpNHY8iwn8KyIAzm1KBdtRZWErpI5lEWTXp5Z/vHpZ3/wyKKwYGGOdAYwR0EZwoezTYApBEIObyELl/aE1/83cp40Pt5mxqCKrE4Ck+mVWKKcI5tA8BLEhRBKJLjez6a7MLq7XZtp+yyOawwCBtkiBVZDKzRk4NN7NQBMYPHiZDFhXY+p9ff7F961vVcnl4R5I2ykJ5XFN7Ab7Gc61VoipNBKF+PDyztu5lfrSLT/wIwCxq0CAGtXHZTzqR2jtwQiXONma6hHpj9sLT7YaPxfTXuZdBGA02Wi7FS48YiTfj+i2NhqtdhP5RC8mh2/Op7y0v6eAcWVLFT8D7kWX5S9mepp+C450MV6aWL1cGnvkxbwHtLW2B9AOkLeUd9KEDuh9fl/7CEj7YH5g+3r/lWfF9In7tPz6T4IIwBJOr1SJyIGQMZQbsh5P9uBq5VJtqHh2mo49pdw5WFoEwKWqWHacaWOjQXWGcifKo6vj5RGS6zykI587XeUIQDqJSmAp+lE4qt19W5P9o8+Lma5DcjsC8JiT607lMVkdqQ0Vyh3lHhmh52tfNy78ajXv0rgYzv8nfwswANuk+7sD/Q0aAAAAAElFTkSuQmCC">
        <style>
            html, body, pre {
                margin: 0;
                padding: 0;
                font-family: Monaco, 'Lucida Console', monospace;
                background: #ECECEC;
            }
            h1 {
                margin: 0;
                background: #A31012;
                padding: 20px 45px;
                color: #fff;
                text-shadow: 1px 1px 1px rgba(0,0,0,.3);
                border-bottom: 1px solid #690000;
                font-size: 28px;
            }
            p#detail {
                margin: 0;
                padding: 15px 45px;
                background: #F5A0A0;
                border-top: 4px solid #D36D6D;
                color: #730000;
                text-shadow: 1px 1px 1px rgba(255,255,255,.3);
                font-size: 14px;
                border-bottom: 1px solid #BA7A7A;
            }
        </style>
    </head>
    <body>
        <h1>Oops, an error occurred</h1>

        <p id="detail">
            This exception has been logged with id <strong>7ccin52ja</strong>.
        </p>

    </body>
</html>
```

Here are my Java candidates:

```bash
$ tree -L 1 .sdkman/candidates/java/
.sdkman/candidates/java/
├── 11.0.3.hs-adpt
├── 11.0.3.j9-adpt
├── 11.0.3-zulu -> java-11-hotspot
├── 12.0.1.hs-adpt
├── 12.0.1.j9-adpt
├── 19.0.2-grl
├── 7.0.222-zulu
├── 8.0.212.hs-adpt
├── 8.0.212.j9-adpt
├── current -> /home/vpavic/.sdkman/candidates/java/11.0.3.hs-adpt
├── java-11-hotspot -> 11.0.3.hs-adpt
├── java-11-openj9 -> 11.0.3.j9-adpt
├── java-12-hotspot -> 12.0.1.hs-adpt
├── java-12-openj9 -> 12.0.1.j9-adpt
├── java-7-hotspot -> 7.0.222-zulu
├── java-8-hotspot -> 8.0.212.hs-adpt
└── java-8-openj9 -> 8.0.212.j9-adpt
```

**Comment by darek1024** [comment]
@vpavic the same happens for me, I want to install Java 12 but even don't know which versions are available 😔

**Comment by marc0der** [comment]
Just for the record, this was due to some faulty data in the database that caused the rendering of the Java list view to misbehave. I've also put some other safeguards in place. @vpavic can you confirm that everything is working for you again?

**Comment by vpavic** [comment]
Yes @marc0der, I can confirm it's all good now. :+1:

**Comment by tsabirgaliev** [comment]
This would also let me register jdk home symlinks in my IDE and update process wouldn't  break my IDE workflow

**Comment by rgoldberg** [comment]
`<vendor>` should be renamed `<dist>` (short for distribution).

`<dist>` support should be general (i.e. not tied to `java`), so that it can be used when necessary for other candidates.

if `<dist>` is left out for a candidate that supports dists, then sdkman can use some multi-distribution upgrade order similar to what it does now for `java` (though the current ordering for java doesn't make sense, so it should be improved).

`sdk dist` should list all the candidates that support dists.

`sdk dist <candidate>` should list all the dists supported by `<candidate>`

If you don't want to redo command line parsing by supporting an extra `<dist>` argument, you could support dists by appending `-` then `<dist>` to `<candidate>`, e.g.:

`sdk <command> [<candidate>[-<dist>]]`

Instead of having `<dist>` as a separate argument, e.g.:

`sdk <command> [<candidate> [<dist>]]`

I like the idea of multi-axis distribution qualifiers, though maybe they should be unordered labels rather than hierarchical.  If hierarchical, you'd have to remember their ordering, plus you couldn't have cross-cutting concerns.  e.g., say multiple distributions use the same jvm, and you just care about the latest version of that jvm, not which vendor has packaged it up (I don't know if j9 is available outside of AdoptOpenJDK, but if it is, j9 & hs could both be used by multiple vendors).

Given that labels aren't hierarchical, a syntax like `a,b,c` is probably preferable to `a/b/c`.

In the code, I'd assign each label to an axis, e.g. j9 & hs would both be values for the jvm field for the java candidate.  I'd also allow multiple values for the same field to be synonymous (e.g., hs & hotspot) so people can use either more descriptive values or terser ones (plus you can generate auto-documentation by showing that hs = hotspot).  Every value should be unique across all values for all fields for the same candidate.  e.g., hs couldn't be a value for 2 different fields for the same candidate.  This will allow specifying labels without their field name, though a field name should be optional, like `[<field>=]value`, so `jvm=hs` & `hs` would be equivalent.

`*` would be a special non-unique value that could only be used after the field name, e.g.: `fx=*`, which would indicate that the fx label must be present for the dist, but the exact value doesn't matter (which specifies in this case that the java candidate must support JavaFX, but the type of the JavaFX support doesn't matter).

Boolean fields would have the field name be the same as the only value for that field, e.g., long-term support could be requested via `lts=lts` or just `lts`.

Example fields would be: vendor (e.g., oracle, azul, amazon), jvm (e.g., hs, j9, graalvm), fx (e.g., openjfx), lts (lts), etc.  name would be a field where each dist is required to have a unique value (whereas more than one dist can be from the same vendor, or use the same jvm, or supply JavaFX, etc.).

So, e.g., if you just want long-term-support versions from openjdk, you could use `sdk i java open,lts`.

If no dist matches all labels, or if an optional version was specified, but that version doesn't match the specified labels (e.g., `sdk i java open,lts 13.0.1`), then sdkman should fail with an error.

Allowing specifying on the command line only one value per field would allow sdkman to report a more descriptive error for a situation like `a,b,c,j9,d,e,f,hs,g,h,i`, because it could say that you must specify at most one of j9 & hs, instead of just saying that no distribution matches `a,b,c,j9,d,e,f,hs,g,h,i`.

As an addendum, I think that you should also support locking version prefixes so that `sdk upgrade` only offers updates with the same prefix.  e.g., If version `1.2.3` of a candidate is installed, if I just want upgrades for 1.x (not 2.x, 3.x, etc.), then lock @ `1`; so, `1.2.3.1`, `1.2.4` & `1.3` would all be allowed as upgrades, but `2.0` wouldn't.  If I want upgrades for 1.2.x (but not 1.3.x, etc.), then lock @ `1.2`; so, `1.2.3.1` & `1.2.4` would both be allowed as upgrades, but `1.3` & `2.0` wouldn't.  You should be able to specify multiple locks for a candidate, so, e.g., if I have 11.0.5, 12.0.2 & 13.0.1 , I could lock at 11, 12.0 & 13.0.1 to vary the prefixes per version.  The syntax for locking could be:

`sdk lock <candidate> [dist] <comma-separated-version-prefixes>`

e.g.,:

`sdk lock java open 11,12.0,13.0.1`

I can open a new issue for version prefix locking, but I figure if you implement dists, you might be refactoring a lot, and knowing about version prefix locking beforehand might allow you to implement dists in a way that makes it easier to add version prefix locking later.

**Comment by ksilz** [comment]
I have JDK 11 in the **OpenJ9** flavor installed. SDKMAN wants to upgrade this to the **Hotspot** version. I consider this a bug. Does this issue cover "my bug"?

![image](https://user-images.githubusercontent.com/1061209/72208600-5a22e400-349c-11ea-90fc-173f334e0760.png)

**Comment by tsabirgaliev** [comment]
FWIW, I switched to Macports OpenJDK builds. Pros: simultaneous up-to-date OpenJDK 8/11/13 discoverable on Macos by `/usr/libexec/java_home`. Cons: only OpenJDK.

**Comment by marc0der** [comment]
This isn't really a bug, since sdkman only holds a the notion of a default version. In other words, if you aren't on the default version it will always suggest that one as the preferred version.

The reason is that we shoehorned java (with multiple versions of the same candidate) into how all other candidates currently work. 

Once I've reworked how we handle java, these issues should go away.

**Comment by rgoldberg** [comment]
@marc0der Thanks for looking into this. Any thoughts about my proposal above? I know it was very long. Are there any points I could clarify? Are there any parts that seem good, bad, unnecessary, etc.? Thanks again.

**Comment by marc0der** [comment]
@rgoldberg Sorry for not getting back to you before regarding your proposal. I'll be very honest to say that I simply haven't gotten around to going through all your points yet.

My suggestion would be to break the proposal down into smaller, more digestible bits and then raise PRs where you see fit. I'll be honest to say that I barely have time to implement existing feature requests (all done in my spare time), let alone to refactor the entire codebase.

This project was built from the start through a series of small incremental changes, and has so evolved into what it is today. It's not perfect, but it is solid and makes many people's lives easier. It isn't merely a bash script, but it spans a datastore, many microservices and an extensive bash client. Making such huge changes in the underlying domain will take up much time, time which I simply don't have available.

All that said, if you feel strongly enough about your proposal and are willing to put some action behind your words (and of course if I and the others agree with what is proposed), then I would be happy to merge your PRs. I'm also always available on slack, feel free to drop me a note in the cli-development channel if you would like to chat about individual proposal points prior to doing the work.

Hope that helps!

**Comment by rgoldberg** [comment]
@marc0der Sorry for late reply.

I'm not a bash expert, but I might be able to look into this at some point.

I'm happy to converse via Slack.

I agree that work should be broken down into manageable chunks, but I think that it would be beneficial to have a completely envisioned UX & code refactoring outcome before working on this.  I think it would make sense to discuss any UX revamps holistically before looking at the code, then figure out the requisite code refactoring, to determine the effort required for each UX change as well as their interdependencies.

**Comment by spencergibb** [comment]
Not sure if this is the right place to comment or if I should open a new issue. I'm looking for something similar for the `sdk env` usage with `.sdkmanrc`. I'd like to say I want `java=8,hs-adpt` (ie java 8 from adopt openjdk, or 11 or 13 or whatever major) without worrying about the minor versions. Then I can check that into source control for a project

**Comment by DennisBecker** [comment]
I also have problems with the upgrade handling. ALso, I wanted to upgrade the java 8 version of Zulu to a newer java 8 version of Zulu, not any version 11, as I have version 11 and 8 installed from Zulu.

**Comment by Karasiq** [comment]
Same. I need a pattern like "8-zulu-fx", because with strict minor version specified, i often get errors like "Stop! java 8.0.262.fx-zulu is not available" in my builds because java was updated and the old version removed.

**Comment by woj-tek** [comment]
I was meaning to open new FR but luckily found this one - I think it would be sensible to split both vendor and handling of the major version (or maybe even some sort of notion of "LTS", but that would complicate things to much I think). IMHO quite often people would like to install some major version of JDK from particular vendor, so something along the lines of @marc0der suggested here:
> extend the base commands to include a vendor qualifier, so allowing something like sdk install java [vendor] [version]. it would also work for the set and default commands. we might need to think this through a bit more.

but with the consideration that `[version]` would be either major version (so `sdk install java adpt 11` and it would automatically upgrade all minor versions) or particular, specific version (i.e. `sdk install java adpt 11.0.8.hs`)

**Comment by hisnameishb** [comment]
Some missing features:

* turn off the sdk-version for a specific topic and any topic.
That means the sdk installed variant is not usable. Usefull if you want to use a other vairant
that is manually installed or comes with your linux distro.
sdk off java :  turns off only sdk-installed java.

If you know perlbrew, nodebrew, phpbrew,... they all have such a switch.

* List installed sdk software
Sometimes I forgot all the stuff I installed with sdk.
You can do a $SDKDIR/candidates  or go with a filemager to this dir and check it out
but that feels wrong.
Btw: If you you rm all versions of a major package there is still an empty dir in candidates,
this should be removed also.


I want something like this:
  sdk list installed
shows all top level entries in this dir. Empty dirs sholud be removed.

**Comment by rgoldberg** [comment]
I could provide a PR for this issue, along with adding a complementary pinning functionality, as long as a decision is made about what should be done. I would suggest:

1. The concept of vendor should be renamed distribution since a vendor is an organization, like Adopt OpenJDK. Adopt OpenJDK supplies 2 different distributions, HotSpot & J9.

2. Whatever is done for `java` should be made generic to work with any candidate that will have multiple distributions.

3. Let's call a candidate that has multiple distributions (like `java`) a "multi-distribution candidate", "mdc" for brevity (if anyone has a better term, please suggest it, as I don't love this term). An mdc will not have any versions directly associated with it. Instead, each of its distributions will have its own list of versions; upgrading a distribution will only possibly offer a different version of that same distribution, not any version of any other distribution for the same (or any other) mdc.

4. Each mdc could have a default distribution, and each distribution would have its own default version, so `sdk i java` would still work without having to specify a distribution.

5. Each distribution would be referenced on the command-line (and elsewhere) with a single consecutive non-white-space string, starting with the mdc name, followed by a dash, followed by the distribution part. e.g., `java-amzn`. Keeping the mdc & candidate as one string allows existing (and new) commands to parse arguments in the same manner, avoids variable argument counts, and allows distributions to be used in commands that accept a list of candidates. `-` should not be allowed in normal candidate names; it would be reserved as the mdc / distribution separator. (Some other character could be used as the separator, if people prefer a different one)

6. Distribution names would be removed from versions. So, it would be something like distribution `java-amzn` & version `16.0.1` instead of the existing `java` & `16.0.1-amzn`, respectively.

7. A new command, `sdk distributions <mdc>`, would list all the distributions for `<mdc>`.

8. Existing commands like `sdk list java` would list all versions for all distributions for `java`. `sdk list java-amzn` would list all versions of `java-amzn`, and no versions of other distributions or candidates. 

9. LTS versions could be handled many ways. A simple way is to append `-lts` to the version number for each LTS version.

10. A new command, `sdk pin <candidate> [<version-regex>] <pin-mode>` would be added, where `<pin-mode>` could be `none`, `beta`, `milestone`, `rc`, `stable`, `lts`, `major`, `minor`, `patch`, or a custom regex (the last can be omitted, if you have problems with it). The default pin-mode would be `lts`. `none` upgrades to the latest, including prereleases. `beta`, `milestone`, `rc`, `stable` & `lts` upgrade to the latest version with the applicable lifecycle marker (or a lifecycle marker that is more stable; e.g., an RC marker is more stable than `beta`) in its version name. `major`, `minor` & `patch` would upgrade only if the version component matched an already installed version. e.g., if you have `gradle` `6.8.2` & `5.7.0` installed, and you pin it to `major`, then if `7.0` is released, nothing will be upgraded. If, however, `6.9.0` is released, sdkman will offer to upgrade from`6.8.2`. Pin mode could also be a combination of one of the three semantic version segment pin modes (`major`, `minor` & `patch`) with one of the other pin modes, separated with a dash, e.g., `major-beta` would upgrade to newer versions with the same major version, as long as the version is a beta or more stable. `major` would be the equivalent of `major-stable`. To support ad hoc semantic version levels, `sv<level>` pin modes could be supported, where e.g., `sv1` is equivalent to `major`, `sv2` is `minor`, etc.

11. The current concept of a default version to install could just be replaced by either installing the latest LTS version, or, if there are no LTS versions, the latest stable version. The default pin mode could be `LTS`. Maybe a pin mode argument could be added to `sdk install`, to simplify proper pinning, and so that new users learn about pin modes when they learn how to use install, so people aren't confused when they're only offered LTS upgrades.

(NB: I'm willing to clarify and/or modify any part of this proposal if people have questions and/or issues with it)

**Comment by marc0der** [comment]
@rgoldberg We won't be making such broad sweeping changes to SDKMAN at this stage. If you feel the need for so many changes, please feel free to fork the project and apply them to your own fork. We will keep working on our current roadmap to deliver small incremental improvements with the time that we have available to us.

**Comment by gabrieljones** [comment]
I have the following situation
```
 Temurin       |     | 18           | tem     |            | 18-tem              
               |     | 18.0.1       | tem     |            | 18.0.1-tem          
               |     | 17.0.3       | tem     |            | 17.0.3-tem          
               | >>> | 17.0.2       | tem     | installed  | 17.0.2-tem          
               |     | 11.0.15      | tem     |            | 11.0.15-tem         
               |     | 11.0.14      | tem     |            | 11.0.14-tem         
               |     | 11.0.14.1    | tem     | installed  | 11.0.14.1-tem       
               |     | 8.0.332      | tem     |            | 8.0.332-tem         
               |     | 8.0.322      | tem     |            | 8.0.322-tem   
```
Is there any easy way to upgrade to the latest 17-tem minor version `17.0.3-tem` and the latest 11-tem minor version `11.0.15-tem` without having to do
```
sdk install java 17.0.3-tem
sdk install java 11.0.15-tem
sdk uninstall java 17.0.2-tem
sdk uninstall java 11.0.14.1-tem 
```

Can I just do `sdk upgrade java` and have all that done for me?

**Comment by whitebear60** [comment]
I have installed java 17.0.3-ms and 8.332.08.1-amzn (the fallback one, if some older software will not run on Java 17). When I'm trying `sdk upgrade`, it forces me to install java `17.0.3-tem`, which I don't want to.

The similar case is for grade. I've installed the latest rc (7.5-rc-2), however, upgrade prompts me to downgrade to 7.4.2.
I think that this behavior needs to be changed...
Not sure if to open the new issue regarding it, or it will be considered duplicate and I need to write about this here...

**Comment by BuckB** [comment]
> The expectations could be expressed as:
> 
>     * given I have a JDK from vendor X installed, when `upgrade` command is used SDKMAN should offer to install a new release from vendor X (if exists)
> 
>     * given I have a JDK from vendor X installed, when `upgrade` command is used SDKMAN shouldn't offer to install a release from vendor Y (or any other vendor that isn't X)

This! I would consider this a top-tier improvement. You can set a remote default in case the user hasn't changed it locally, but check if it equals the user's local default and run the command against the user settings. If I run sdk upgrade it runs only on the remote default vendor. I have to list the vendor I want and check it. It goes against the moto: "Simplifying life. No more hunting for downloads" when the user is still hunting for downloads.

**Comment by marc0der** [comment]
I'm still working hard and making good progress with the underlying plumbing. The backend is nearing the cutover point, so I can proceed with the data migration to Postgres (yes! it goes that deep!). After that, I can start the actual work of porting the remainder of the CLI to Rust, which will solve this problem. It's sad that you're not with us anymore, @woj-tek, although I understand your frustration 😞
